pkg/app/pipedv1/plugin/scriptrun/plugin.go

```diff
@@ -81,27 +81,66 @@ func (p *plugin) ExecuteStage(ctx context.Context, _ sdk.ConfigNone, _ sdk.Deplo
 	switch input.Request.StageName {
 	case stageScriptRun:
 		return &sdk.ExecuteStageResponse{
-			Status: executeScriptRun(ctx, input.Request, input.Client.LogPersister()),
+			Status: executeScriptRun(ctx, input.Request, input.Client.LogPersister(), input.Client),
 		}, nil
 	case stageScriptRunRollback:
-		panic("unimplemented")
+		return &sdk.ExecuteStageResponse{
+			Status: executeRollback(ctx, input.Request, input.Client.LogPersister(), input.Client),
+		}, nil
 	}
 	return nil, fmt.Errorf("unsupported stage %s", input.Request.StageName)
 }
 
-func executeScriptRun(ctx context.Context, request sdk.ExecuteStageRequest[struct{}], lp sdk.StageLogPersister) sdk.StageStatus {
+func executeScriptRun(ctx context.Context, request sdk.ExecuteStageRequest[struct{}], lp sdk.StageLogPersister, metadataStore deploymentMetadataStore) sdk.StageStatus {
 	lp.Infof("Start executing the script run stage")
 	opts, err := decode(request.StageConfig)
 	if err != nil {
 		lp.Errorf("failed to decode the stage config: %v", err)
 		return sdk.StageStatusFailure
 	}
+	if err = metadataStore.PutDeploymentPluginMetadata(ctx, strconv.Itoa(request.StageIndex), "_"); err != nil {
+		lp.Errorf("failed to mark stage as run: %v", err)
+		return sdk.StageStatusFailure
+	}
 	if opts.Run == "" {
 		return sdk.StageStatusSuccess
 	}
 	c := make(chan sdk.StageStatus, 1)
 	go func() {
-		c <- executeCommand(opts, request, lp)
+		c <- executeCommand(opts.Run, opts.Env, request, lp)
+	}()
+	select {
+	case result := <-c:
+		return result
+	case <-ctx.Done():
+		lp.Info("ScriptRun cancelled")
+		// We can return any status here because the piped handles this case as cancelled by a user,
+		// ignoring the result from a plugin.
+		return sdk.StageStatusFailure
+	}
+}
+func executeRollback(ctx context.Context, request sdk.ExecuteStageRequest[struct{}], lp sdk.StageLogPersister, metadataStore deploymentMetadataStore) sdk.StageStatus {
+	lp.Infof("Start executing the script run rollback stage")
+	opts, err := decode(request.StageConfig)
+	if err != nil {
+		lp.Errorf("failed to decode the stage config: %v", err)
+		return sdk.StageStatusFailure
+	}
+	_, err = metadataStore.GetDeploymentPluginMetadata(ctx, strconv.Itoa(request.StageIndex))
+	if strings.HasPrefix(err.Error(), "metadata store not found") {
+		lp.Infof("skip rollback for stage %s as it hasn't run", request.StageName)
```

Comment:
Yours will be `skip rollback for stage SCRIPT_RUN_ROLLBACK as it hasn't run`.

```suggestion
		lp.Infof("skip this rollback stage because the SCRIPT_RUN stage of index %d has not run", request.StageIndex)
```